### PR TITLE
finish commit

### PR DESCRIPTION
**Pull request by sergeynakul**

### CODE REVIEW

**Review by KonstantinFaleev — APPROVED** [review]
Ok!

**Review by derunva — APPROVED** [review]
Nice!

File: db/seed.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module Seed
```
Comment (epoberezhny) [inline db/seed.rb]: Make it a separate class please

File: moduls/validation.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        name = validate[:variable].capitalize
+        send("validate_#{validate[:type]}", value, name, *validate[:args])
+      end
+    end
+
+    def validate_presence(value, name)
+      raise PresenceError, "#{name} can't be blank" if value.to_s.empty? || value.nil?
+    end
+
+    def validate_positive(value, name)
```
Comment (epoberezhny) [inline moduls/validation.rb]: Consider using of validation of type "greatest or equal"

File: models/author.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class Author
```
Comment (epoberezhny) [inline models/author.rb:3]: Maybe we should add some getters?

File: models/book.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class Book
+  include Validation
+
+  attr_reader :title
```
Comment (epoberezhny) [inline models/book.rb]: reader for author?

File: models/library.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+class Library
+  include Seed
+
+  BOOK_QUANTITY = 3
+  DATA_FILE = 'db/data.yml'
+
+  attr_accessor :authors, :books, :readers, :orders
```
Comment (epoberezhny) [inline models/library.rb]: attr_reader

File: models/library.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  def top_readers(quantity)
+    most_popular(quantity, :reader)
+  end
+
+  def top_books(quantity)
+    most_popular(quantity, :book)
+  end
+
+  def number_readers_of_top_books
```
Comment (epoberezhny) [inline models/library.rb]: Should have an argument

File: models/library.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    @orders.select { |order| top_books(BOOK_QUANTITY).include? order.book }.map(&:reader).uniq.length
+  end
+
+  private
+
+  def most_popular(quantity, entity)
+    @orders.group_by(&entity).max_by(quantity) { |_, value| value.length }.to_h.keys
```
Comment (epoberezhny) [inline models/library.rb]: map(&:first)

File: models/library.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  def top_readers(quantity)
+    most_popular(quantity, :reader)
+  end
+
+  def top_books(quantity)
+    most_popular(quantity, :book)
+  end
+
+  def number_readers_of_top_books
+    @orders.select { |order| top_books(BOOK_QUANTITY).include? order.book }.map(&:reader).uniq.length
```
Comment (epoberezhny) [inline models/library.rb]:
```ruby
popular_books = top_books(BOOK_QUANTITY)
@orders.select { |order| popular_books.include? order.book }.map(&:reader).uniq.length
```

**Review by epoberezhny — CHANGES_REQUESTED** [review]

File: moduls/validation.rb (hunk 2 above)
Comment (sergeynakul) [reply moduls/validation.rb]: remind, please what did you mean? I must rename method validate_positive?

File: models/library.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+class Library
+  DATA_FILE = 'db/data.yml'
+
+  attr_reader :authors, :books, :readers, :orders
```
Comment (epoberezhny) [inline models/library.rb]: Remove it please

File: db/seed.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+class Seed
+  class << self
```
Comment (epoberezhny) [inline db/seed.rb]: Make it a 'classic' class

**Comment by epoberezhny** [comment]
Add ruby version please

**Review by epoberezhny — CHANGES_REQUESTED** [review]

File: index.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+require_relative './models/order'
+require_relative './errors/presence_error'
+require_relative './errors/type_class_error'
+require_relative './errors/number_value_error'
+
+library = Library.new
+seed = Seed.new(library)
+seed.generate_data
+library.store_data
+library.load_data
```
Comment (artem247) [inline index.rb:21]: It seems like seeded data is never assigned back to the original library on which statistics calculations are performed

Comment (epoberezhny) [reply index.rb:21]: What do you mean?

File: models/library.rb (hunk 9 above)
Comment (epoberezhny) [reply models/library.rb]: What about this?

**Review by epoberezhny — CHANGES_REQUESTED** [review]

File: index.rb (hunk 11 above)
Comment (sergeynakul) [reply index.rb:21]: 1. library = Library.new - create instance of Library with @authors = [ ], @books = [ ], @readers = [ ], @orders = [ ]
2. seed = Seed.new(library) - create instance of Seed
3. seed.generate_data - fill library @authors, @books, @readers, @orders with faker data
4. library.store_data - save library data in yaml file
5. library.load_data - read library data from yaml file

Comment (epoberezhny) [reply index.rb:21]: Ok. But you can create a new instance of library and load data

**Review by epoberezhny — APPROVED** [review]

**Pull request merged**